Update go-control-plane to v0.9.0

thomasco opened the pull request:
This pulls in new protos from envoy via envoyproxy/go-control-plane. Additionally, there was a switch from istio/gogo-genproto to golang/protobuf for code generation in go-control-plane.

One of the proto changes is the [inclusion of the metadata_context field in service.auth.v2.AttributeContext](https://www.envoyproxy.io/docs/envoy/latest/api-v2/service/auth/v2/attribute_context.proto.html?highlight=metadata_context), which can be used to share JWT context with OPA (see metadata_context_namespaces in [ext_authz](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/ext_authz/v2/ext_authz.proto.html))

ashutosh-narkar commented on go.mod:
```diff
@@ -27,10 +26,11 @@ require (
+	google.golang.org/grpc v1.23.0
 	gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127 // indirect
 	gopkg.in/fsnotify.v1 v1.4.7 // indirect
+	istio.io/gogo-genproto v0.0.0-20190614210408-e88dc8b0e4db // indirect
```
I wonder what depends on `istio.io/gogo-genproto`  🤔

ashutosh-narkar commented:
@thomasco these changes look good ! It would be helpful to update the README with an example of the `metadata_context ` field, an updated [input](https://github.com/open-policy-agent/opa-istio-plugin#example-input), policy etc.

thomasco commented:
Rebased onto master and added docs. Sorry I've been putting this off for so long.

Probably more of a go convention question, but can I hand-edit go.mod to remove the remaining gogo stuff? `go mod tidy` removes gogo/googleapis and gogo-genproto, but also a bunch of golang.org/x and other things.

ashutosh-narkar commented:
> Rebased onto master and added docs. Sorry I've been putting this off for so long.
> 
> Probably more of a go convention question, but can I hand-edit go.mod to remove the remaining gogo stuff? `go mod tidy` removes gogo/googleapis and gogo-genproto, but also a bunch of golang.org/x and other things.

Yes @thomasco we need to fix that. I can work on that in a separate PR. It's ok to leave it as-is for now.

ashutosh-narkar commented on README.md:
````diff
@@ -346,6 +346,79 @@ allow = response {
+    metadata_context_namespaces:
+    - envoy.filters.http.jwt_authn
+    <...>
+```
````
Is this configuration applicable to the `EnvoyFilter` kind  ? Or does it need to be updated ?

ashutosh-narkar commented on README.md:
````diff
@@ -346,6 +346,79 @@ allow = response {
+}
+```
+
+### Example OPA Configuration
````
Nit: `Example OPA Policy`

ashutosh-narkar commented on README.md:
````diff
@@ -346,6 +346,79 @@ allow = response {
+```
+
+### Example OPA Configuration
+This JWT data can be accessed in OPA configuration like this:
````
Nit: `configuration ` -> `policy`

ashutosh-narkar commented:
@thomasco the changes look fine. Just a clarification on how we can make this applicable to the `EnvoyFilter` kind.

danielpops commented:
> @thomasco the changes look fine. Just a clarification on how we can make this applicable to the `EnvoyFilter` kind.

@ashutosh-narkar, @thomasco and I don't have any familiarity with istio (assuming that when you say `EnvoyFilter`, you're referring to https://github.com/open-policy-agent/opa-istio-plugin/blob/master/quick_start.yaml#L5 ?). We can provide what *might* be the right config, which I think would look like 

```
diff --git a/quick_start.yaml b/quick_start.yaml
index 53f387a7..ecf01894 100644
--- a/quick_start.yaml
+++ b/quick_start.yaml
@@ -40,6 +40,20 @@ spec:
         google_grpc:
           target_uri: 127.0.0.1:9191
           stat_prefix: "ext_authz"
+      metadata_context_namespaces:
+      - envoy.filters.http.jwt_authn
+  - insertPosition:
+      index: FIRST
+    listenerMatch:
+      listenerType: SIDECAR_INBOUND
+      listenerProtocol: HTTP
+    filterType: HTTP
+    filterName: "envoy.filters.http.jwt_authn"
+    filterConfig:
+      providers:
+        example:
+          payload_in_metadata: verified_jwt
+
 ---
 ############################################################
 # Namespace for cluster-wide OPA-Istio components.
```

But it will be difficult for us to test.

ashutosh-narkar commented:
> > @thomasco the changes look fine. Just a clarification on how we can make this applicable to the `EnvoyFilter` kind.
> 
> @ashutosh-narkar, @thomasco and I don't have any familiarity with istio (assuming that when you say `EnvoyFilter`, you're referring to https://github.com/open-policy-agent/opa-istio-plugin/blob/master/quick_start.yaml#L5 ?). We can provide what _might_ be the right config, which I think would look like
> 
> ```
> diff --git a/quick_start.yaml b/quick_start.yaml
> index 53f387a7..ecf01894 100644
> --- a/quick_start.yaml
> +++ b/quick_start.yaml
> @@ -40,6 +40,20 @@ spec:
>          google_grpc:
>            target_uri: 127.0.0.1:9191
>            stat_prefix: "ext_authz"
> +      metadata_context_namespaces:
> +      - envoy.filters.http.jwt_authn
> +  - insertPosition:
> +      index: FIRST
> +    listenerMatch:
> +      listenerType: SIDECAR_INBOUND
> +      listenerProtocol: HTTP
> +    filterType: HTTP
> +    filterName: "envoy.filters.http.jwt_authn"
> +    filterConfig:
> +      providers:
> +        example:
> +          payload_in_metadata: verified_jwt
> +
>  ---
>  ############################################################
>  # Namespace for cluster-wide OPA-Istio components.
> ```
> 
> But it will be difficult for us to test.

@danielpops this looks accurate to me. We can merge this and work on any README updates later as needed. Thanks @thomasco for these changes.
